File: daily/0328/l-1022 Sum of Root To Leaf Binary Numbers.cpp
#include <iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    int res = 0;

    void helper(TreeNode* root, int num) {
        if (!root) return;

        num = num * 2 + root->val;
        if (!root->left && !root->right) res += num;
        else {
            if (root->left) helper(root->left, num);
            if (root->right) helper(root->right, num);
        }
    }

    int helper2(TreeNode* root, int prev) {
        if (!root) return 0;
        int cur = 2 * prev + root->val;
        return root->left == root->right ? cur : helper2(root->left, cur) + helper2(root->right, cur);
    }

public:
    int sumRootToLeaf(TreeNode* root) {
        helper(root, 0);
        return res;
    }

    int woRes(TreeNode* root, int num) {
        return helper2(root, 0);
    }
};

File: daily/march/0328/l-1022 Sum of Root To Leaf Binary Numbers.py
import sys

sys.path.append("C:\\Users\\didme654\\Documents\\visual studio code\\algorithms_v2\\daily\\freq_methods")

import listToTree

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.res = 0

    def sumRootToLeaf(self, root: TreeNode) -> int:
        self.helper(root, 0)
        return self.res
        
    def helper(self, root: TreeNode, prev: int) -> None:
        cur = 2 * prev + root.val
        if not root.left and not root.right:
            self.res += cur
        else:
            if root.left:
                self.helper(root.left, cur)
            if root.right:
                self.helper(root.right, cur)


if __name__ == "__main__":
    S = Solution()
    tree = [1,0,1,0,1,0,1]
    root = listToTree.listToTree(tree)
    print(S.sumRootToLeaf(root))